test/Microsoft.Azure.Devices.Edge.Test.Common/LeafDevice.cs

```diff
@@ -0,0 +1,352 @@
+    {
+        readonly DeviceClient client;
+        readonly Device device;
+        readonly IotHub iotHub;
+        readonly string messageId;
+
+        LeafDevice(Device device, DeviceClient client, IotHub iotHub)
+        {
+            this.client = client;
+            this.device = device;
+            this.iotHub = iotHub;
+            this.messageId = Guid.NewGuid().ToString();
+        }
+
+        public static Task<LeafDevice> CreateAsync(
+            string leafDeviceId,
+            Protocol protocol,
+            AuthenticationType auth,
+            Option<string> parentId,
+            bool useSecondaryCertificate,
+            EdgeCertificateAuthority edgeCa,
+            IotHub iotHub,
+            CancellationToken token)
+        {
+            return Profiler.Run(
+                async () =>
+                {
+                    ITransportSettings transport = protocol.ToTransportSettings();
+                    Platform.InstallEdgeCertificates(edgeCa.Certificates.TrustedCertificates, transport);
+
+                    string edgeHostname = Dns.GetHostName().ToLower();
+
+                    switch (auth)
+                    {
+                        case AuthenticationType.Sas:
+                            return await CreateWithSasAsync(
+                                leafDeviceId,
+                                parentId,
+                                iotHub,
+                                transport,
+                                edgeHostname,
+                                token);
+
+                        case AuthenticationType.CertificateAuthority:
+                            {
+                                string p = parentId.Expect(() => new ArgumentException());
+                                return await CreateWithCaCertAsync(
+                                    leafDeviceId,
+                                    p,
+                                    edgeCa,
+                                    iotHub,
+                                    transport,
+                                    edgeHostname,
+                                    token);
+                            }
+
+                        case AuthenticationType.SelfSigned:
+                            {
+                                string p = parentId.Expect(() => new ArgumentException());
+                                return await CreateWithSelfSignedCertAsync(
+                                    leafDeviceId,
+                                    p,
+                                    useSecondaryCertificate,
+                                    edgeCa,
+                                    iotHub,
+                                    transport,
+                                    edgeHostname,
+                                    token);
+                            }
+
+                        default:
+                            throw new InvalidEnumArgumentException();
+                    }
```

Comment:
In existing E2E test script, there are 8 tests running in Transparent gateway tests; which are "LeafDevice with SAS auth not in scope", "LeafDevice with SAS auth in scope", "x509CA" and "x509Thumprint".   Where is x509Thumprint? is it equivalent to SelfSigned authentication type?

Reply:
Yes, self-signed is thumbprint.